app/js/lib/foxbox/network.js

```diff
@@ -0,0 +1,206 @@
+    return `${this.settings.localScheme}://${this.settings.localHostname}:${this.settings.localPort}`;
+  }
+
+  get tunnelOrigin() {
+    return `${this.settings.tunnelScheme}://${this.settings.tunnelHostname}:${this.settings.tunnelPort}`;
+  }
+
+  get connected() {
+    return this[p.local] || this[p.remote];
+  }
+
+  get connection() {
+    if (this[p.local]) {
+      return 'local';
+    } else if (this[p.remote]) {
+      return 'remote';
+    }
+    return 'unknown';
+  }
+
+  /**
+   * Request a JSON from a specified URL.
+   *
+   * @param {string} url The URL to send the request to.
+   * @param {string} method The HTTP method (defaults to "GET").
+   * @param {Object} body An object of key/value.
+   * @return {Promise}
+   */
+  fetchJSON(url, method = 'GET', body = undefined) {
+    return this[p.fetch](url, 'application/json', method, body)
+      .then(response => response.json());
+  }
+
+  /**
+   * Request a Blob from a specified URL.
+   *
+   * @param {string} url The URL to send the request to.
+   * @param {string} blobType The Blob mime type (eg. image/jpeg).
+   * @param {string} method The HTTP method (defaults to "GET").
+   * @param {Object} body An object of key/value.
+   * @return {Promise<Blob>}
+   */
+  fetchBlob(url, blobType, method, body) {
+    return this[p.fetch](url, blobType, method, body)
+      .then(response => response.blob());
+  }
+
+  /**
+   * Request a content of the specified type from a specified URL.
+   *
+   * @todo Detect if the URL is relative, if so prepend this.origin.
+   *
+   * @param {string} url The URL to send the request to.
+   * @param {string} accept The content mime type (eg. image/jpeg).
+   * @param {string=} method The HTTP method (defaults to "GET").
+   * @param {Object=} body An object of key/value.
+   * @return {Promise}
+   * @private
+   */
+  [p.fetch](url, accept, method = 'GET', body = undefined) {
+    method = method.toUpperCase();
+
+    const req = {
+      method,
+      headers: { Accept: accept },
+      cache: 'no-store'
+    };
+
+    if (method === 'POST' || method === 'PUT') {
+      req.headers['Content-Type'] = 'application/json;charset=UTF-8';
+    }
+
+    if (this.settings.session) {
+      // The user is logged in, we authenticate the request.
+      req.headers.Authorization = `Bearer ${this.settings.session}`;
+    }
+
+    if (body !== undefined) {
+      req.body = JSON.stringify(body);
+    }
+
+    return fetch(url, req)
+      .then(res => {
+        if (!res.ok) {
+          throw new TypeError(
+            `The response returned a ${res.status} HTTP status code.`
+          );
+        }
+
+        return res;
+      })
+      .catch(error => {
+        console.error('Error occurred while fetching content: ', error);
+        throw error;
+      });
+  }
+
+  /**
+   * Ping the box to detect whether we connect locally or remotely.
+   *
+   * @private
+   */
+  [p.pingBox]() {
+    this[p.ping](`${this.localOrigin}/ping`)
```

Comment:
> I don't a think remote connection can ever be faster than local network. Am I wrong?

I think it can happen (admit that very rarely though), just any local network glitch and remote will be faster :) But anyway it's not related to this patch